core/src/main/java/com/netflix/conductor/core/execution/tasks/SystemTaskExecutor.java

```diff
@@ -0,0 +1,114 @@
+
+    private final int callbackTime;
+    private final QueueDAO queueDAO;
+
+    ExecutionConfig defaultExecutionConfig;
+    private final WorkflowExecutor workflowExecutor;
+    private final Configuration config;
+
+    ConcurrentHashMap<String, ExecutionConfig> queueExecutionConfigMap = new ConcurrentHashMap<>();
+
+    SystemTaskExecutor(QueueDAO queueDAO, WorkflowExecutor workflowExecutor, Configuration config) {
+        this.config = config;
+        int threadCount = config.getIntProperty("workflow.system.task.worker.thread.count", 10);
+        this.callbackTime = config.getIntProperty("workflow.system.task.worker.callback.seconds", 30);
+
+        String threadNameFormat = "system-task-worker-%d";
+        this.defaultExecutionConfig = new ExecutionConfig(threadCount, threadNameFormat);
+        this.workflowExecutor = workflowExecutor;
+        this.queueDAO = queueDAO;
+
+        LOGGER.info("Initialized the SystemTaskExecutor with {} threads and callback time: {} seconds", threadCount,
+            callbackTime);
+    }
+
+    void pollAndExecute(String queueName) {
+        // get the remaining capacity of worker queue to prevent queue full exception
+        ExecutionConfig executionConfig = getExecutionConfig(queueName);
+        SemaphoreUtil semaphoreUtil = executionConfig.getSemaphoreUtil();
+        ExecutorService executorService = executionConfig.getExecutorService();
+        String taskName = QueueUtils.getTaskType(queueName);
+
+        if (!semaphoreUtil.canProcess()) {
+            // no available permits, do not poll
+            return;
+        }
+        try {
+            List<String> polledTaskIds = queueDAO.pop(queueName, 1, 200);
+            Monitors.recordTaskPoll(queueName);
+            LOGGER.debug("Polling queue:{}, got {} tasks", queueName, polledTaskIds.size());
+            if (polledTaskIds.size() == 1 && StringUtils.isNotBlank(polledTaskIds.get(0))) {
+                String taskId = polledTaskIds.get(0);
+                LOGGER.debug("Task: {} from queue: {} being sent to the workflow executor", taskId, queueName);
+                Monitors.recordTaskPollCount(queueName, "", 1);
+
+                WorkflowSystemTask systemTask = SystemTaskWorkerCoordinator.taskNameWorkflowTaskMapping.get(taskName);
+                CompletableFuture<Void> taskCompletableFuture = CompletableFuture.runAsync(() ->
+                    workflowExecutor.executeSystemTask(systemTask, taskId, callbackTime), executorService);
+
+                // release permit after processing is complete
+                taskCompletableFuture.whenComplete((r, e) -> semaphoreUtil.completeProcessing());
+            } else {
+                // no task polled, release permit
+                semaphoreUtil.completeProcessing();
+            }
+        } catch (Exception e) {
+            // release the permit if exception is thrown during polling, because the thread would not be busy
+            semaphoreUtil.completeProcessing();
```

Comment:
`semaphoreUtil.completeProcessing();` this can be part of finally block.

Reply:
If this is put in a finally block, permits will be released twice in case of successful system task execution, once at line 88 (after the future is completed) and then again in finally block.